### PR TITLE
[Fix #2812] Bash solution to see five number summary on spent time on PR or on opened PR

### LINKED ISSUE #2812: Bash script that allows one to see a five-number summary of PR open time
### User Story

As a developer, I want to know how long PRs stay open over time so that I can improve the code delivery speed.

### Description

*Type*: Feature/Tooling

*Summary*: Write a bash script that allows one to see a five-number summary of PR open time.

`./scripts/pr-statistics.sh` => 

```
Min: ...
Q1: ...
Mid: ...
Q3: ...
Max: ...
```

Additionally, the following options don't have to be exposed as arguments but should be changeable in the script:

- repo
- start time and end time
- print Q3 PR URLs

### Notes
https://en.wikipedia.org/wiki/Five-number_summary

### PR DESCRIPTION
**Pull request by StudioEtrange**
Fixes #2812 

### Summary:

Bash solution to see five number summary of PR.
Take care of multi-pages list of PR.
Can change github repo by fixing variable `GITHUB_REPO`

### Testing notes (optional):
Tested on regular MacOSX and linux (Debian)

### Steps to test:

```
./scripts/pr-statistics.sh -h
```

status: ready

### CODE REVIEW

**Comment by StudioEtrange** [comment]
Two differents type of solution implemented, depends on what you want to see

## Solution 1 output pr-statistics.sh

Print creation date of only opened PR

> Tested on https://github.com/status-im/status-react
> Min: lundi 20 novembre 2017, 23:28:45 (UTC+0100)
> Q1: vendredi 22 décembre 2017, 13:17:41 (UTC+0100)
> Mean: jeudi 28 décembre 2017, 17:00:31 (UTC+0100)
> Q3: dimanche 31 décembre 2017, 22:13:41 (UTC+0100)
> Max: mercredi 3 janvier 2018, 22:19:03 (UTC+0100)

NOTE : output is in french here cause of my system, on other system, it will output local date.


## Solution 2 output  pr-statistics-spent-time.sh

Print amount of time spent on closed PR.
As github API use 30 PR max by page, it could be slow when there is a lot of closed PRs

> Tested on https://github.com/status-im/status-dev-cli
> Min: 12
> Q1: 22656
> Mean: 245305
> Q3: 7546970
> Max: 10459664

**Comment by oskarth** [comment]
Thanks for the contribution @StudioEtrange! I'll take a look at this soon.

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: Can we default to status-react?

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: Can we print a progress bar? like

[Fetching PR lists: 1/75]
[Fetching PR lists: 2/75]

etc.

Also, are there any timeout/GH API block concerns?

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: What does this do exactly? Can we turn it off by default? When I run script I see:

DEBUG -- url is https://api.github.com/repos/status-im/status-dev-cli/pulls?state=closed -- nb page : 1
DEBUG -- downloading page 1 ...
DEBUG -- page 1 downloaded
Min: 12
Q1: 22656
Mean: 245305
Q3: 7546970
Max: 10459664

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: Can we print this? so if I run script with it says something like:

[Fetching PR stats from status-react]

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: Can we convert this from seconds to something more human readable? Like DD HH MM

File: pr-statistics-spent-time.sh
Comment (oskarth) [inline pr-statistics-spent-time.sh]: Can we format this by columns? So all first columns take up X width and the numbers are aligned

File: scripts/pr-statistics.sh
Comment (oskarth) [inline scripts/pr-statistics.sh]: Column format here too, can we also add days open? Something like

Mon Nov 20 22:28:45 UTC 2017 (45 days)

**Review by oskarth — CHANGES_REQUESTED** [review]
@StudioEtrange Good start! Made some comments

**Comment by oskarth** [comment]
@StudioEtrange Oh right, I realize this wasn't super clear in initial issue, but what I meant by:

> start time and end time

Is that one should be able to run `./scripts/pr-statistics.sh` with some variable like:
```
START_TIME=2017-11-01
END_TIME=2017-12-01
```
And then script should only check PRs in that interval.

The use case is to track how we improve merge speed over time.

**Comment by StudioEtrange** [comment]
code reviewed with your requests

- merged two compute mode (open and merge PR) in only one pr-statistics.sh file
- add optional start end time
- add usage
- add verbose mode
- print results in fixed columns way
- print results in human readable way

### About computational mode

- open : catch only open PR, not yet merged nor closed and then compute five number summary on opened time until today (DEFAULT MODE)
- merge : catch only merged AND closed PR and then compute five number summary on time spent before PR was merged

### About github api limitation

When requesting PR, we are limited to max 30 PR by request. So depending of start_time end_time  and chosen mode (open, merge), it could be a lof of PR and requests

### How to test

```
./scripts/pr-statistics.sh -h
./scripts/pr-statistics.sh
./scripts/pr-statistics.sh open
./scripts/pr-statistics.sh merge
```

  ### Samples of output from github status-im/status-react

- **stats on open PRs**
- verbose mode active
- with START_TIME="2017-01-01" and END_TIME="2017-12-01"
```
[Fetching open PR stats from status-im/status-react]
[Analysing PR lists: 1/1]
[Stats computed on 30 values]
             Min |               Q1 |             Mean |               Q3 |              Max
  0d 00h:36m:08s |   0d 23h:45m:24s |   6d 11h:34m:26s |  10d 20h:55m:20s |  44d 21h:43m:29s
```

- **stats on already merged PRs**
- verbose mode active
- with START_TIME="2017-01-01" and END_TIME="2017-12-01"
```
[Analysing PR lists: 34/41]
[Stats computed on 15224 values]
             Min |               Q1 |             Mean |               Q3 |              Max
  0d 00h:00m:07s |   0d 00h:52m:31s |   0d 16h:07m:08s |   3d 06h:50m:58s |  71d 01h:44m:14s
```

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: changed to a variable named VERBOSE. It allow to print some info while computing.
done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

There is no timeout concerns, but github api allow us to get PR by batch of 30 max only

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

File: pr-statistics-spent-time.sh
Comment (StudioEtrange) [reply pr-statistics-spent-time.sh]: done in commit https://github.com/status-im/status-react/pull/2893/commits/0b287d778877586a7249552f78af30f2223027ac

**Comment by oskarth** [comment]
Thanks, looks good to me! One final comment from me: can we print START and END time to STDOUT? Easy mistake to make otherwise.

**Review by oskarth — APPROVED** [review]

**Comment by StudioEtrange** [comment]
Ok i add print out of date time.
And then I rebase PR

**Comment by StudioEtrange** [comment]
Ok
* rebase done
* print out selected time
* fix a bug with end_time
  

Good to merge :)

**Comment by oskarth** [comment]
Thanks @StudioEtrange! Could you squash commits to one please? And then just need a second reviewer to sanity check, and we can merge after they OK it. Pinged internal chat.

**Comment by StudioEtrange** [comment]
rebase and squash done !

**Comment by StudioEtrange** [comment]
Any news ? 
Since last time where everything was ok, now there is a jenkis warn. But I cannot access logs of the build and know why

**Comment by andytudhope** [comment]
Our apologies @StudioEtrange - this is release week for us, so this issue was moved down the priority list. Am sure @oskarth will come back to it, likely early next week 👍 Thanks again for the great work!

**Comment by oskarth** [comment]
@StudioEtrange Couldn't find another reviewer, but since I asked for this, am happy with solution and it doesn't impact rest of project, I'm going to merge it now. Thanks and sorry about the delay!

**Pull request merged**
